ui/src/core_components/layout/CoreStep.vue

```diff
@@ -0,0 +1,405 @@
+ */
+
+const STEP_BIT_INSTANCE_NUMBER = 0;
+const CONTENT_DISPLAYING_INSTANCE_NUMBER = 1;
+
+import { Component, FieldType, InstancePath } from "../../streamsyncTypes";
+import { useEvaluator } from "../../renderer/useEvaluator";
+import {
+	contentHAlign,
+	cssClasses,
+	contentPadding,
+} from "../../renderer/sharedStyleFields";
+import { onBeforeUnmount } from "vue";
+
+const description =
+	"A container component that displays its child components as a step inside a Step Container.";
+
+export default {
+	streamsync: {
+		name: "Step",
+		description,
+		allowedParentTypes: ["steps", "repeater"],
+		allowedChildrenTypes: ["*"],
+		category: "Layout",
+		fields: {
+			name: {
+				name: "Name",
+				default: "(No name)",
+				init: "Step name",
+				type: FieldType.Text,
+			},
+			contentPadding: {
+				...contentPadding,
+				default: "16px",
+			},
+			isCompleted: {
+				name: "Completed",
+				desc: "Use a state element to mark this step as complete.",
+				default: "no",
+				type: FieldType.Text,
+				options: {
+					yes: "Yes",
+					no: "No",
+				},
+			},
+			contentHAlign,
+			cssClasses,
+		},
+		previewField: "name",
+	},
+};
+</script>
+<script setup lang="ts">
+import { Ref, computed, inject, onBeforeMount, watch } from "vue";
+import injectionKeys from "../../injectionKeys";
+import BaseContainer from "./../base/BaseContainer.vue";
+import { type StepsData } from "./CoreSteps.vue";
+
+const fields = inject(injectionKeys.evaluatedFields);
+const instancePath = inject(injectionKeys.instancePath);
+const instanceData = inject(injectionKeys.instanceData);
+const isBeingEdited = inject(injectionKeys.isBeingEdited);
+
+const ss = inject(injectionKeys.core);
+const ssbm = inject(injectionKeys.builderManager);
+const componentId = inject(injectionKeys.componentId);
+const { isComponentVisible } = useEvaluator(ss);
+const selectedId = computed(() => ssbm?.getSelectedId());
+
+const stepContainerData: Ref<StepsData> = getStepContainerData();
+
+/**
+ * Returns the InstanceData for the parent Step Container.
+ */
+function getStepContainerData(): Ref<StepsData> {
+	for (let i = -1; i > -1 * instancePath.length; i--) {
+		const item = instancePath.at(i);
+		const component = ss.getComponentById(item.componentId);
+		if (!component) return;
+		const type = component.type;
+		if (type !== "steps") continue;
+		return instanceData.at(i);
+	}
+	return;
+}
+
+/**
+ * Returns the distance between this Step instance and the child of the parent Container, starting from the former.
+ * The direct child isn't necessarily a Step, it can be a Repeater.
+ */
+function getDirectChildInstanceNegativeIndex() {
+	for (let i = -2; i > -1 * instancePath.length; i--) {
+		const item = instancePath.at(i);
+		const component = ss.getComponentById(item.componentId);
+		if (!component) return;
+		const type = component?.type;
+		if (type !== "steps") continue;
+		return i + 1;
+	}
+	return;
```

Comment:
nitpick: those 2 functions are almost identical, maybe extract function like `getContainerIndex` or something like this and calculate it just once.
question: Will the step component always be a direct child of the steps component?

Reply:
Good point about repetition, I'll refactor, which I'm hoping will help with clarity too.

No, the _Step_ won't always be a direct child of the _Step Container_ (`steps`). That's what brings the complexity and what makes this harder. If that was the case I could get the parent as easily as `instancePath.at(-2)` rather than in this convoluted way.